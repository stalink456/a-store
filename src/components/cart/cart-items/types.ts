export enum OptionItems {
  Color = 'цвет',
  Size = 'размер',
  stickerName = 'номер стикера',
}
