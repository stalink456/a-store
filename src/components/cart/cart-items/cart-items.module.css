@import url("../../../css/variables.css");

.cart__items__item {
  display: flex;
  justify-content: space-between;

  padding: 10px 0;
}

.item__counter__buttons {
  display: flex;
  flex-direction: row;
}

.item__wrapper {
  display: flex;
  flex-direction: row;
}

.item__wrapper__description p {
  margin: 2px 0;
}

.cart__items__item img {
  width: 70px;
  height: 70px;

  border-radius: 5px;
  object-fit: cover;
  object-position: top 20%;
}

.item__counter__buttons_button span {
  width: 25px;
}

.item__counter__buttons_button,
.item__counter__buttons_button svg {
  width: 25px;
  height: 30px;
}

.item__counter__buttons_cross,
.item__counter__buttons_cross svg {
  width: 25px;
  height: 30px;
}

.item__counter__buttons_button:active,
.item__counter__buttons_cross:active {
  transform: scale(1.1);
}

.item__counter__buttons__text_count input {
  width: 20px;
  padding: 0;

  text-align: center;

  background-color: var(--white-color);
}

.item__counter__buttons__text_count input:hover {
  border: 1px solid var(--grey-color);
}

.item__text_price {
  width: 80px;
  margin: 0;

  text-align: center;
}

.cart__items__empty_cart {
  margin: 10px 0;

  font-size: 24px;
}

@media (width <= 620px) {
  .cart__items__item {
    position: relative;
  }

  .item__counter__buttons_cross {
    position: absolute;
    top: 0;
    right: 0;
  }

  .cart__items__item {
    align-items: flex-start;
    flex-direction: column;
    justify-content: flex-start;
  }

  .item__counter__buttons_cross svg {
    color: var(--logo-color);
  }
}

@media (width <= 395px) {
  .item__counter__buttons_cross {
    right: -10px;
  }
}
