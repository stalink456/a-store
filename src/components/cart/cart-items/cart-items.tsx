import React from 'react';
import { Space } from '@alfalab/core-components/space';
import { Typography } from '@alfalab/core-components/typography';
import { MaskedInput } from '@alfalab/core-components/masked-input';
import { LoadableImage } from 'components/ui-components/loadable-image';
import { HiOutlinePlusCircle, HiOutlineMinusCircle } from 'react-icons/hi';
import { RxCrossCircled } from 'react-icons/rx';
import { IconButton } from '@alfalab/core-components/icon-button';
import { useCart } from 'hooks/use-cart';
import { OptionItems } from './types';

import styles from './cart-items.module.css';

const mask = { input: [/^[1-9]/, /[0-9]*$/] };

export const CartItems: React.FC = React.memo(() => {
  const {
    cartItemsLength,
    cartItems,
    onClickIncrement,
    onClickDecrement,
    onClickRemove,
    handleChangeCount,
  } = useCart();

  const renderOption = (name: OptionItems, option: string | null) => {
    return option ? (
      <Typography.Text tag='p' weight='regular' view='secondary-large'>
        {name}: {option}
      </Typography.Text>
    ) : null;
  };

  const renderCartItems = () => {
    return cartItemsLength ? (
      cartItems.map(
        ({
          id,
          preview,
          title,
          totalCount,
          totalPrice,
          color,
          size,
          stickerName,
        }) => (
          <Space
            direction='horizontal'
            key={id}
            align='center'
            className={styles.cart__items__item}
          >
            <Space direction='horizontal' align='center'>
              <LoadableImage src={preview} alt='test' />
              <Space className={styles.item__description} size={0}>
                <Typography.Text tag='p' weight='bold' view='primary-large'>
                  {title}
                </Typography.Text>
                {renderOption(OptionItems.Color, color)}
                {renderOption(OptionItems.Size, size)}
                {renderOption(OptionItems.stickerName, stickerName)}
              </Space>
            </Space>

            <Space
              className={styles.item__counter}
              fullWidth
              size={0}
              direction='horizontal'
              align='center'
            >
              <Space
                direction='horizontal'
                align='center'
                className={styles.item__counter__buttons}
                size={5}
                fullWidth
              >
                <IconButton
                  className={styles.item__counter__buttons_button}
                  icon={HiOutlineMinusCircle}
                  size='s'
                  disabled={totalCount === 1}
                  onClick={() => onClickDecrement(id)}
                  data-testid='decrement-count'
                />

                <MaskedInput
                  value={totalCount.toString()}
                  defaultValue={totalCount.toString()}
                  onChange={(e) => handleChangeCount(e, id)}
                  placeholder={totalCount.toString()}
                  mask={mask.input}
                  className={styles.item__counter__buttons__text_count}
                  data-testid='input-count'
                />

                <IconButton
                  className={styles.item__counter__buttons_button}
                  icon={HiOutlinePlusCircle}
                  size='s'
                  disabled={totalCount === 99}
                  onClick={() => onClickIncrement(id)}
                  data-testid='increment-count'
                />

                <Typography.Text
                  weight='bold'
                  tag='p'
                  className={styles.item__text_price}
                >
                  {totalPrice} ₽
                </Typography.Text>
              </Space>
              <IconButton
                className={styles.item__counter__buttons_cross}
                icon={RxCrossCircled}
                size='s'
                onClick={() => onClickRemove(id)}
                data-testid='delete-item'
              />
            </Space>
          </Space>
        )
      )
    ) : (
      <Typography.Text
        className={styles.cart__items__empty_cart}
        tag='p'
        weight='bold'
        view='primary-large'
      >
        Корзина пуста
      </Typography.Text>
    );
  };

  return <div className={styles.cart__items}>{renderCartItems()}</div>;
});
