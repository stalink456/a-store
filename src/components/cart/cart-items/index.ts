export { CartItems } from "./cart-items";
