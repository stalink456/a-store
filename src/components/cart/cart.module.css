@import url("../../css/variables.css");
@import url("../../css/animation.css");

.cart__wrapper {
  position: fixed;
  z-index: 1;
  top: 100px;
  right: 20px;

  border-radius: 50px;
  background-color: var(--logo-color);
  composes: fade_scale from global;
}

.cart__icon {
  padding: 20px;

  color: var(--white-color);

  font-size: 54px;
}

.cart__icon_text {
  position: absolute;
  top: 70%;
  right: 0;

  width: 35px;
  padding: 5px;

  text-align: center;

  color: var(--white-color);
  border-radius: 25px;
  background-color: var(--black-color);
}

.cart__icon_tooltip {
  position: absolute;
  top: 25%;
  right: 60%;

  display: flex;
  visibility: hidden;
  align-items: center;
  justify-content: center;

  width: 120px;
  height: 40px;

  transition: opacity 1s ease-in-out;
  transform: translate(-50%);

  opacity: 0;
  color: var(--white-color);
  border-radius: 10px;
  background-color: var(--black-color);
  composes: tooltip_visible from global;
}

.cart__icon_tooltip::before {
  position: absolute;
  top: 50%;
  left: 110%;

  content: "";
  transform: translate(-50%, -50%);

  border: 15px solid;
  border-color: var(--border-color) var(--border-color) var(--border-color)
    var(--black-color);
}

.cart__icon_tooltip__text {
  margin: 0;
}

.cart__wrapper:hover .cart__icon_tooltip {
  visibility: visible;

  opacity: 1;
}

.cart__wrapper:hover {
  transform: scale(1.1);
}

@media (width <= 960px) {
  .cart__wrapper {
    top: 80vh;
  }
}

@media (width <= 640px) {
  .cart__icon {
    padding: 10px;

    font-size: 40px;
  }

  .cart__icon_text {
    top: 60%;
    left: 60%;

    width: 30px;
    padding: 3px;
  }

  .cart__icon_tooltip {
    top: 20%;
    right: 40%;
  }
}
