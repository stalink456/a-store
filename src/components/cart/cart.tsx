import React from "react";
import { IconButton } from "@alfalab/core-components/icon-button";
import { AiOutlineShopping } from "react-icons/ai";
import { Typography } from "@alfalab/core-components/typography";
import { Space } from "@alfalab/core-components/space";
import { useSidePanel } from "hooks/use-side-panel";
import { CartPanel } from "./cart-panel/cart-panel";
import { useCart } from "hooks/use-cart";

import styles from "./cart.module.css";

export const Cart: React.FC = React.memo(() => {
  const { cartItemsLength, totalCount, cartTotalPrice, cartRef } = useCart();
  const { open, handleOpenMenu } = useSidePanel(cartRef);

  const renderCart = () => {
    return cartItemsLength ? (
      <div key={totalCount} className={styles.cart__wrapper}>
        <IconButton
          className={styles.cart__icon}
          icon={AiOutlineShopping}
          size="s"
          onClick={handleOpenMenu}
          data-testid="cart-icon"
        />
        <Typography.Text className={styles.cart__icon_text} tag="p">
          {totalCount}
        </Typography.Text>
        <Space className={styles.cart__icon_tooltip}>
          <Typography.Text
            className={styles.cart__icon_tooltip__text}
            tag="p"
            weight="bold"
            view="primary-large"
          >
            = {cartTotalPrice} â‚½
          </Typography.Text>
        </Space>
      </div>
    ) : null;
  };

  return (
    <div className={styles.cart}>
      {renderCart()}
      <CartPanel ref={cartRef} open={open} handleOpenMenu={handleOpenMenu} />
    </div>
  );
});
