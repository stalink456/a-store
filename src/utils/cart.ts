import { CartType } from "types";

export const calcTotalPrice = (items: CartType[]) => {
  return items.reduce((sum, obj) => {
    return obj.price * obj.totalCount + sum;
  }, 0);
};

export const getCartFromLS = () => {
  const data = localStorage.getItem("cart");
  const items: CartType[] = data ? JSON.parse(data) : [];
  const cartTotalPrice = calcTotalPrice(items);

  return {
    items: items,
    cartTotalPrice,
  };
};

export const calculateTotalCount = (items: CartType[]) => {
  return items.reduce((sum: number, { totalCount }) => sum + totalCount, 0);
};

export const generateID = (
  color: string | null,
  size: string | null,
  stickerName: string | null
) => {
  return [color, size, stickerName].filter((value) => value !== null).join("");
};
