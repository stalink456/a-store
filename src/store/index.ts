import { configureStore } from '@reduxjs/toolkit';
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import createSagaMiddleware from 'redux-saga';
import { madeInAlfaReducer } from './made-in-alfa';
import { yourDesignReducer } from './your-design';
import { productReducer } from './product';
import { cartReducer } from './cart';
import { orderReducer } from './order';
import { notificationsReducer } from './notifications';
import { rootSaga } from './root-saga';

const sagaMiddleware = createSagaMiddleware();
const middlewares = [sagaMiddleware];

export const store = configureStore({
  reducer: {
    madeInAlfa: madeInAlfaReducer,
    yourDesign: yourDesignReducer,
    product: productReducer,
    cart: cartReducer,
    order: orderReducer,
    notifications: notificationsReducer,
  },
  devTools: true,
  middleware: middlewares,
});

sagaMiddleware.run(rootSaga);

export type ApplicationState = ReturnType<typeof store.getState>;
export const useAppDispatch = () => useDispatch<typeof store.dispatch>();
export const useAppSelector: TypedUseSelectorHook<ApplicationState> =
  useSelector;
