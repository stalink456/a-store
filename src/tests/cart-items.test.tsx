import { MemoryRouter } from "react-router-dom";
import { fireEvent, render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import "@testing-library/jest-dom";
import { CartItems } from "components/cart/cart-items";
import { store } from "store";
import { cartActions } from "store/cart";
import * as cart from "./mocks/get-cart-item-200.json";
import * as confirm from "./mocks/confirm-window-200.json";

const mockedCart = cart["cart"];
const mockedConfirm = confirm["confirm"];

describe("CartItems component", () => {
  beforeAll(() => store.dispatch(cartActions.addItem(mockedCart)));

  it("should render CartItems component", async () => {
    const { baseElement } = render(
      <Provider store={store}>
        <CartItems />
      </Provider>,
      { wrapper: MemoryRouter }
    );

    expect(baseElement).toMatchSnapshot();
  });

  it("should render CartItems component with name of product", async () => {
    render(
      <Provider store={store}>
        <CartItems />
      </Provider>,
      { wrapper: MemoryRouter }
    );

    const result = await screen.findByText("Рюкзак «Для умных и свободных»");
    expect(result).toBeInTheDocument();
  });

  it("should render CartItems component with increment count", async () => {
    render(
      <Provider store={store}>
        <CartItems />
      </Provider>,
      { wrapper: MemoryRouter }
    );

    expect(screen.getByTestId("decrement-count")).toBeDisabled();

    expect(screen.getByTestId("input-count")).toHaveValue("1");

    fireEvent.click(screen.getByTestId("increment-count"));
    expect(screen.queryByTestId("input-count")).toHaveValue("2");
    expect(screen.getByTestId("decrement-count")).not.toBeDisabled();
  });

  it("should render CartItems component with decriment count", async () => {
    render(
      <Provider store={store}>
        <CartItems />
      </Provider>,
      { wrapper: MemoryRouter }
    );

    expect(screen.getByTestId("decrement-count")).not.toBeDisabled();

    fireEvent.click(screen.getByTestId("decrement-count"));

    expect(screen.getByTestId("decrement-count")).toBeDisabled();
  });

  it("should render CartItems component with delete item", async () => {
    window.confirm = jest.fn().mockImplementation(() => mockedConfirm);

    render(
      <Provider store={store}>
        <CartItems />
      </Provider>,
      { wrapper: MemoryRouter }
    );

    expect(screen.getByTestId("input-count")).toHaveValue("1");

    fireEvent.click(screen.getByTestId("delete-item"));

    expect(screen.getByText("Корзина пуста")).toBeInTheDocument();
  });

  it("shouldn't render CartItems component with non-existent text", async () => {
    render(
      <Provider store={store}>
        <CartItems />
      </Provider>,
      { wrapper: MemoryRouter }
    );

    const result = screen.queryByText("non-existent text") as HTMLImageElement;
    expect(result).not.toBeTruthy();
  });
});
