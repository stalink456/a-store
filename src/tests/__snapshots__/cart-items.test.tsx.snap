// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItems component should render CartItems component 1`] = `
<body>
  <div>
    <div
      class="cart__items"
    >
      <div
        class="space__spaceContainer_nr3as space__center_nr3as cart__items__item"
      >
        <div
          class="space__spaceItem_nr3as"
          style="margin-right: 16px;"
        >
          <div
            class="space__spaceContainer_nr3as space__center_nr3as"
          >
            <div
              class="space__spaceItem_nr3as"
              style="margin-right: 16px;"
            >
              <div
                class="skeleton__component_1bshy skeleton__animate_1bshy"
              >
                <img
                  alt="test"
                  class=""
                  src="http://qa-games.ru/astore/public/images/15932051.jpeg"
                />
              </div>
            </div>
            <div
              class="space__spaceItem_nr3as"
            >
              <div
                class="space__spaceContainer_nr3as space__vertical_nr3as space__start_nr3as item__description"
              >
                <div
                  class="space__spaceItem_nr3as"
                >
                  <p
                    class="typography__paragraph_c16y5 typography__primary-large_c16y5 typography__bold_c16y5"
                  >
                    Рюкзак «Для умных и свободных»
                  </p>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="space__spaceItem_nr3as"
        >
          <div
            class="space__spaceContainer_nr3as space__center_nr3as space__spaceContainerFullWidth_nr3as item__counter"
          >
            <div
              class="space__spaceItem_nr3as space__spaceItemFullWidth_nr3as"
              style="margin-right: 0px;"
            >
              <div
                class="space__spaceContainer_nr3as space__center_nr3as space__spaceContainerFullWidth_nr3as item__counter__buttons"
              >
                <div
                  class="space__spaceItem_nr3as space__spaceItemFullWidth_nr3as"
                  style="margin-right: 5px;"
                >
                  <button
                    class="button__component_j4291 button__ghost_j4291 button__s_j4291 button__component_3xn9f button__ghost_3xn9f item__counter__buttons_button icon-button__primary_2t17d icon-button__component_2t17d"
                    data-testid="decrement-count"
                    disabled=""
                    type="button"
                  >
                    <span
                      class="button__text_j4291 button__stretchText_j4291"
                    >
                      <span
                        class="icon-button__iconWrapper_pctl3 icon-button__s_pctl3"
                      >
                        <svg
                          class="icon-button__icon_pctl3"
                          fill="none"
                          height="1em"
                          stroke="currentColor"
                          stroke-width="0"
                          viewBox="0 0 24 24"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M15 12H9m12 0a9 9 0 11-18 0 9 9 0 0118 0z"
                            stroke-linecap="round"
                            stroke-linejoin="round"
                            stroke-width="2"
                          />
                        </svg>
                      </span>
                    </span>
                  </button>
                </div>
                <div
                  class="space__spaceItem_nr3as space__spaceItemFullWidth_nr3as"
                  style="margin-right: 5px;"
                >
                  <div
                    class="form-control__component_e23cr form-control__component_1xedm item__counter__buttons__text_count form-control__s_e23cr"
                  >
                    <div
                      class="form-control__inner_e23cr form-control__inner_1xedm form-control__filled_e23cr form-control__filled_1xedm"
                    >
                      <div
                        class="form-control__inputWrapper_e23cr"
                      >
                        <div
                          class="form-control__input_e23cr"
                        >
                          <input
                            class="input__input_96c2k input__input_js93g"
                            data-testid="input-count"
                            placeholder="1"
                            type="text"
                            value="1"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
                <div
                  class="space__spaceItem_nr3as space__spaceItemFullWidth_nr3as"
                  style="margin-right: 5px;"
                >
                  <button
                    class="button__component_j4291 button__ghost_j4291 button__s_j4291 button__component_3xn9f button__ghost_3xn9f item__counter__buttons_button icon-button__primary_2t17d icon-button__component_2t17d"
                    data-testid="increment-count"
                    type="button"
                  >
                    <span
                      class="button__text_j4291 button__stretchText_j4291"
                    >
                      <span
                        class="icon-button__iconWrapper_pctl3 icon-button__s_pctl3"
                      >
                        <svg
                          class="icon-button__icon_pctl3"
                          fill="none"
                          height="1em"
                          stroke="currentColor"
                          stroke-width="0"
                          viewBox="0 0 24 24"
                          width="1em"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <path
                            d="M12 9v3m0 0v3m0-3h3m-3 0H9m12 0a9 9 0 11-18 0 9 9 0 0118 0z"
                            stroke-linecap="round"
                            stroke-linejoin="round"
                            stroke-width="2"
                          />
                        </svg>
                      </span>
                    </span>
                  </button>
                </div>
                <div
                  class="space__spaceItem_nr3as space__spaceItemFullWidth_nr3as"
                >
                  <p
                    class="typography__paragraph_c16y5 item__text_price typography__primary-medium_c16y5 typography__bold_c16y5"
                  >
                    4999
                     ₽
                  </p>
                </div>
              </div>
            </div>
            <div
              class="space__spaceItem_nr3as space__spaceItemFullWidth_nr3as"
            >
              <button
                class="button__component_j4291 button__ghost_j4291 button__s_j4291 button__component_3xn9f button__ghost_3xn9f item__counter__buttons_cross icon-button__primary_2t17d icon-button__component_2t17d"
                data-testid="delete-item"
                type="button"
              >
                <span
                  class="button__text_j4291 button__stretchText_j4291"
                >
                  <span
                    class="icon-button__iconWrapper_pctl3 icon-button__s_pctl3"
                  >
                    <svg
                      class="icon-button__icon_pctl3"
                      fill="none"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 15 15"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        clip-rule="evenodd"
                        d="M0.877075 7.49988C0.877075 3.84219 3.84222 0.877045 7.49991 0.877045C11.1576 0.877045 14.1227 3.84219 14.1227 7.49988C14.1227 11.1575 11.1576 14.1227 7.49991 14.1227C3.84222 14.1227 0.877075 11.1575 0.877075 7.49988ZM7.49991 1.82704C4.36689 1.82704 1.82708 4.36686 1.82708 7.49988C1.82708 10.6329 4.36689 13.1727 7.49991 13.1727C10.6329 13.1727 13.1727 10.6329 13.1727 7.49988C13.1727 4.36686 10.6329 1.82704 7.49991 1.82704ZM9.85358 5.14644C10.0488 5.3417 10.0488 5.65829 9.85358 5.85355L8.20713 7.49999L9.85358 9.14644C10.0488 9.3417 10.0488 9.65829 9.85358 9.85355C9.65832 10.0488 9.34173 10.0488 9.14647 9.85355L7.50002 8.2071L5.85358 9.85355C5.65832 10.0488 5.34173 10.0488 5.14647 9.85355C4.95121 9.65829 4.95121 9.3417 5.14647 9.14644L6.79292 7.49999L5.14647 5.85355C4.95121 5.65829 4.95121 5.3417 5.14647 5.14644C5.34173 4.95118 5.65832 4.95118 5.85358 5.14644L7.50002 6.79289L9.14647 5.14644C9.34173 4.95118 9.65832 4.95118 9.85358 5.14644Z"
                        fill="currentColor"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </span>
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
